hydra-node/src/Hydra/Chain/Direct/Wallet.hs

```diff
@@ -123,38 +129,45 @@ newTinyWallet ::
   NetworkId ->
   -- | Credentials of the wallet.
   (VerificationKey PaymentKey, SigningKey PaymentKey) ->
-  -- Starting point on the chain. From this onward we will receive blocks on 'update'.
-  ChainPoint ->
   -- | A function to query UTxO, pparams, system start and epoch info from the
   -- node. Initially and on demand later.
   ChainQuery IO ->
+  IO (EpochInfo (Either Text)) ->
   IO (TinyWallet IO)
-newTinyWallet tracer networkId (vk, sk) chainPoint queryUTxOEtc = do
-  utxoVar <- newTVarIO =<< queryUTxOEtc (QueryAt chainPoint) address
+newTinyWallet tracer networkId (vk, sk) queryWalletInfo queryEpochInfo = do
+  walletInfoVar <- newTVarIO =<< initialize
   pure
     TinyWallet
-      { getUTxO =
-          (\(u, _, _, _) -> u) <$> readTVar utxoVar
+      { getUTxO = readTVar walletInfoVar <&> walletUTxO
       , sign = Util.signWith sk
       , coverFee = \lookupUTxO partialTx -> do
-          (walletUTxO, pparams, systemStart, epochInfo) <- readTVar utxoVar
+          -- XXX: We should query pparams here. If not, we likely will have
+          -- wrong fee estimation should they change in between.
+          epochInfo <- queryEpochInfo
```

Comment:
Suggestion: Instead of querying `EpochInfo` here, would it not be "better" to pass it as argument to `coverFee`, and run the query at the beginning of the transaction submission process? This would keep the `TinyWallet` "purer" in the sense that it wouldn't need interactions with the chain as those would be handled by its caller.

Reply:
Consequent change here would be to pass everything to `coverFee`: `SystemStart`, `ProtocolParameters`, and `EpochInfo` .. or even better the cardano-api variant of it.

Not passing `queryEpochInfo` to `newTinyWallet`, but to `mkChain` would likely also impact `createMockChain` from the model-spec PR and lead to conflicts.

I would suggest to follow-up with a PR dedicatd to making the `Wallet` use these types and clean up how things are passed.